internal/repositories/rewards.go

```diff
@@ -0,0 +1,218 @@
+
+	rewards, err := models.Rewards(queryMods...).All(ctx, r.pdb.DBS().Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	edges := make([]*gmodel.EarningsEdge, len(rewards))
+	nodes := make([]*gmodel.Earning, len(rewards))
+
+	for i, rw := range rewards {
+		earning := RewardToAPI(*rw)
+		cursor, err := pHelp.EncodeCursor(RewardsCursor{
+			Week:      earning.Week,
+			VehicleID: tokenID,
+		})
+		if err != nil {
+			return nil, err
+		}
+		edges[i] = &gmodel.EarningsEdge{
+			Node:   &earning,
+			Cursor: cursor,
+		}
+
+		nodes[i] = &earning
+	}
+
+	if len(edges) == 0 {
+		return &gmodel.VehicleEarnings{}, nil
+	}
+
+	endCursor, err := pHelp.EncodeCursor(RewardsCursor{
+		Week:      edges[len(edges)-1].Node.Week,
+		VehicleID: tokenID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	startCursor, err := pHelp.EncodeCursor(RewardsCursor{
+		Week:      edges[0].Node.Week,
+		VehicleID: tokenID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	earningsConn := &gmodel.EarningsConnection{
+		TotalCount: stats.TotalCount,
+		PageInfo: &gmodel.PageInfo{
+			EndCursor:   &endCursor,
+			StartCursor: &startCursor,
+		},
+		Edges: edges,
+		Nodes: nodes,
+	}
+
+	return &gmodel.VehicleEarnings{
+		TotalTokens: dbtypes.NullDecimalToInt(stats.TokenSum),
+		History:     earningsConn,
+		VehicleID:   tokenID,
+	}, nil
+}
+
+func (r *Repository) PaginateVehicleEarningsByID(ctx context.Context, vehicleEarnings *gmodel.VehicleEarnings, first *int, after *string, last *int, before *string) (*gmodel.EarningsConnection, error) {
+	pHelp := helpers.PaginationHelper[RewardsCursor]{}
+	limit, err := pHelp.ValidateFirstLast(first, last, maxPageSize)
+	if err != nil {
+		return nil, err
+	}
+
+	if vehicleEarnings == nil || vehicleEarnings.History == nil {
+		return &gmodel.EarningsConnection{}, nil
+	}
+
+	erningsConnectionCopy := *vehicleEarnings.History
+	earningIdx := map[string]int{}
+	for idx, val := range erningsConnectionCopy.Edges {
+		earningIdx[val.Cursor] = idx
+	}
+
+	var edges []*gmodel.EarningsEdge
+	var nodes []*gmodel.Earning
+
+	if after != nil {
+		startFrom := earningIdx[*after]
+
+		if startFrom == len(erningsConnectionCopy.Edges)-1 {
+			return &gmodel.EarningsConnection{
+				PageInfo:   &gmodel.PageInfo{},
+				TotalCount: vehicleEarnings.History.TotalCount,
+			}, nil
+		}
+
+		if startFrom+1 < len(erningsConnectionCopy.Edges) {
+			startFrom += 1
+		}
+
+		erningsConnectionCopy.Edges = erningsConnectionCopy.Edges[startFrom:]
+		erningsConnectionCopy.Nodes = erningsConnectionCopy.Nodes[startFrom:]
+	}
+
+	if before != nil {
+		startFrom := earningIdx[*before]
+
+		if startFrom == 0 {
+			return &gmodel.EarningsConnection{
+				PageInfo:   &gmodel.PageInfo{},
+				TotalCount: vehicleEarnings.History.TotalCount,
+			}, nil
+		}
+
+		erningsConnectionCopy.Edges = erningsConnectionCopy.Edges[:startFrom]
+		erningsConnectionCopy.Nodes = erningsConnectionCopy.Nodes[:startFrom]
+	}
+
+	if before != nil || last != nil {
```

Comment:
I don't think `before != nil` implies that you should reverse.

Reply:
But this is another problem—in the parent where we did the select, did we ever reverse the order in the case where `last` was specified?